os/kernel/pthread/pthread_keycreate.c

```diff
@@ -125,33 +125,38 @@
  *
  ****************************************************************************/
 
-int pthread_key_create(FAR pthread_key_t *key, CODE void (*destructor)(void *))
+int pthread_key_create(FAR pthread_key_t *key, pthread_destructor_t destructor)
 {
-#if CONFIG_NPTHREAD_KEYS > 0
-	FAR struct tcb_s *rtcb = this_task();
-	FAR struct task_group_s *group = rtcb->group;
+	struct pthread_tcb_s *rtcb = (struct pthread_tcb_s *)this_task();
+	struct task_group_s *group = rtcb->cmn.group;
+	int key_index = 0;
 	int ret = EAGAIN;
 
 	DEBUGASSERT(group);
 
 	/* Check if we have exceeded the system-defined number of keys. */
 
-	if (group->tg_nkeys < PTHREAD_KEYS_MAX) {
-		/* Return the key value */
+	while (key_index < PTHREAD_KEYS_MAX) {
+		/* Find not in-use key */
 
-		*key = group->tg_nkeys;
+		if (!group->tg_keys[key_index]) {
+			/* Mark it is in-use */
 
-		/* Increment the count of global keys. */
+			group->tg_keys[key_index] = IN_USE;
 
-		group->tg_nkeys++;
+			/* Save desctructor */
 
-		/* Return success. */
+			rtcb->pthread_data[key_index].destructor = destructor;
 
-		ret = OK;
-	}
+			/* Return the key value */
+
+			*key = key_index;
+
```

Comment:
key is not validated for NULL before assignment.  test case with NULL can cause crash